paddle/framework/scope.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+#include <unordered_map>
+#include <vector>
+#include "paddle/framework/variable.h"
+#include "paddle/utils/Error.h"
+
+namespace paddle {
+namespace framework {
+
+const static Error AlreadyCreated("Variable has already been created");
+
+/**
+ * Scope is an association of a name to Variable. All variables belong to
+ * `Scope`. You need to specify a scope to run a Net, i.e., `net.Run(&scope)`.
+ * One net can run in different scopes and update different variable in the
+ * scope.
+ */
+class Scope {
+ public:
+  Scope() {}
+
+  explicit Scope(const std::shared_ptr<Scope>& scope) : parent_(scope) {}
+
+  ~Scope() {}
+
+  // Create Variable in this Scope. Return error if Variable already been
+  // created.
+  Error __must_check CreateVariable(const std::string& name);
```

Comment:
I checked Caffe2's source code of the similar functionality at [here](https://github.com/caffe2/caffe2/blob/v0.7.0/caffe2/core/workspace.cc#L215):

```cpp
Blob* Workspace::CreateBlob(const string& name) {
  if (HasBlob(name)) {
    VLOG(1) << "Blob " << name << " already exists. Skipping.";
  } else {
    VLOG(1) << "Creating blob " << name;
    blob_map_[name] = unique_ptr<Blob>(new Blob());
  }
  return GetBlob(name);
}
```

I am impressed that Caffe2's authors tried to handle the error as much as they could inside the function definition. Whereas, we simply return an error and use `__must_check` to require users of our functions to handle the error. The best place to handle the error is at where it happens, where we have the most sufficient information to recover, to report, or to fail.

I'd suggest that we use:

```cpp
Variable* CreateVar(const std::string& name);
```